### ISSUE TITLE
array generic functions type vars are not inferred correctly

### ISSUE THREAD

**Issue by yoosiba** [NumberFour/n4js#281]
``` typescript
class A {}
class B {
    @Final
    public a: A;
    constructor(@Spec spec: ~i~this) {}

    static fromA(a: A): B {
        return new B({a: a});
    }

    @Override
    public toString(): string {
        return `[A] with=${ this.a }`;
    }
}

function processAs(arrA: Array<A>): void {
    //XPECT noerrors --> "Couldn't resolve reference to IdentifiableElement 'toString'."
    arrA.map(a => B.fromA(a)).forEach(b => b.toString());//b is of type 'any'
};

function processAsWithWorkaround(arrA: Array<A>): void {
    //first type var of map bounded correctly
    arrA.<B, Array<A>>map(a => B.fromA(a)).forEach(b => b.toString());
};
```

**Comment by yoosiba** [comment]
@mor-n4 pls review if this is a bug or if I have missed somethings

**Comment by JensN4** [comment]
It's a bug, simplified test:

```
class B { foo() {} }
function trans(a): B { return null; }

function processAs(arrA: Array<any>): void {
    // at the moment, Array<any> is inferred XPECT FIXME typeof x ==> Array<B>
    let x = arrA.map(a => trans(a));
    // XPECT FIXME noerrors -->
    x.pop().foo();
};
```

Explicit passing of type arguments "fixes" the problem. But of course, we should infer that.

**Comment by mor-n4** [comment]
There is a work-around available: explicitly declare the return type of the function expression (not nice, I know, but a bit nicer than explicitly declaring the type arguments of the generic method). Based on Jens' example:

```
function processAs(arrA: Array<any>): void {
    // work-around, for the time being:
    arrA.map( (a):B => trans(a) ).forEach( b => b.foo() );
};
```

**Comment by mor-n4** [comment]
This is a known limitation of the return type inference: we decided to not infer the return type of a function from the type of its contained return statements' expressions, for the time being (but it was planned to revisit that at some point). This only leads to the above problem, if there is no type expectation for one of the type variables, as is the case with methods such as #map().

For collecting such use cases where we need better type inference we have IDE-2140. I have added the above example (the simplified version from Jens) there and are thus closing this as duplicate of IDE-2140.
